File: Java/20.java
// https://leetcode.com/problems/valid-parentheses
// Runtime: 1 ms, faster than 99.43% of Java online submissions for Valid Parentheses.
// Memory Usage: 40.3 MB, less than 87.11% of Java online submissions for Valid Parentheses.
// 05/05/2022 20:46
class Solution {
    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<Character>();
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            if (c == ')' && (stack.size() == 0 || stack.pop() != '('))
                return false;
            else if (c == ']' && (stack.size() == 0 || stack.pop() != '['))
                return false;
            else if (c == '}' && (stack.size() == 0 || stack.pop() != '{'))
                return false;
            else if ("({[".indexOf(c) != -1)
                stack.push(c);
        }
        return stack.size() == 0;
    }
}

File: TypeScript/20.ts
// https://leetcode.com/problems/valid-parentheses/
// Runtime: 125 ms, faster than 22.48% of TypeScript online submissions for Valid Parentheses.
// Memory Usage: 45 MB, less than 19.21% of TypeScript online submissions for Valid Parentheses.
// 03/14/2022 10:52
function isValid(s: string): boolean {
    const stack = [];
    for (const c of s) {
        if (['(','{','['].includes(c))
            stack.push(c);
        else if  (((c == ')' && stack[stack.length - 1] == '(')) ||
                  ((c == '}' && stack[stack.length - 1] == '{')) ||
                  ((c == ']' && stack[stack.length - 1] == '[')))
            stack.pop();
        else
            return false;
    }
    return stack.length == 0;
};
